In [1]:
import numpy as np
import pandas as pd

In [2]:
Data = pd.read_csv('Datasets(1)\\uk_road_accident.csv')

In [3]:
import warnings
warnings.filterwarnings('ignore')

In [4]:
Data

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious      5/6/2019  51.506187   
1       200701BS65737           Serious      2/7/2019  51.495029   
2       200701BS66127           Serious    26-08-2019  51.517715   
3       200701BS66128           Serious    16-08-2019  51.495478   
4       200701BS66837            Slight      3/9/2019  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    18-02-2022  57.374005   
660675  201091NM01881            Slight    21-02-2022  57.232273   
660676  201091NM01935            Slight    23-02-2022  57.585044   
660677  201091NM01964           Serious    23-02-2022  57.214898   
660678  201091NM02142           Serious    28-02-2022  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2                      NaN               Urban                    NaN   
3       Single carriageway               Urban     Fine no high winds   
4                      NaN               Urban                    NaN   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  
0                                        Car  
1                                        Car  
2                      Taxi/Private hire car  
3       Bus or coach (17 or more pass seats)  
4                              Other vehicle  
...                                      ...  
660674                                   Car  
660675                                   Car  
660676                                   Car  
660677                 Motorcycle over 500cc  
660

In [5]:
Data.describe()

Latitude      Longitude  Number_of_Casualties  Number_of_Vehicles
count  660654.000000  660653.000000         660679.000000       660679.000000
mean       52.553866      -1.431210              1.357040            1.831255
std         1.406922       1.383330              0.824847            0.715269
min        49.914430      -7.516225              1.000000            1.000000
25%        51.490690      -2.332291              1.000000            1.000000
50%        52.315641      -1.411667              1.000000            2.000000
75%        53.453452      -0.232869              1.000000            2.000000
max        60.757544       1.762010             68.000000           32.000000

<h1>Changing Data types</h1>
<hr>


In [6]:
Data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 660679 entries, 0 to 660678
Data columns (total 14 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   Index                    660679 non-null  object 
 1   Accident_Severity        660679 non-null  object 
 2   Accident Date            660679 non-null  object 
 3   Latitude                 660654 non-null  float64
 4   Light_Conditions         660679 non-null  object 
 5   District Area            660679 non-null  object 
 6   Longitude                660653 non-null  float64
 7   Number_of_Casualties     660679 non-null  int64  
 8   Number_of_Vehicles       660679 non-null  int64  
 9   Road_Surface_Conditions  659953 non-null  object 
 10  Road_Type                656159 non-null  object 
 11  Urban_or_Rural_Area      660664 non-null  object 
 12  Weather_Conditions       646551 non-null  object 
 13  Vehicle_Type             660679 non-null  object 
dtypes: f

In [7]:
Data["Index"] = Data["Index"].astype("category")
Data["Accident_Severity"] = Data["Accident_Severity"].astype("category")
Data["Accident Date"] = Data["Accident Date"].astype("category")
Data["Light_Conditions"] = Data["Light_Conditions"].astype("category")
Data["District Area"] = Data["District Area"].astype("category")
Data["Road_Surface_Conditions"] = Data["Road_Surface_Conditions"].astype("category")
Data["Road_Type"] = Data["Road_Type"].astype("category")
Data["Urban_or_Rural_Area"] = Data["Urban_or_Rural_Area"].astype("category")
Data["Weather_Conditions"] = Data["Weather_Conditions"].astype("category")
Data["Vehicle_Type"] = Data["Vehicle_Type"].astype("category")

In [8]:
Data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 660679 entries, 0 to 660678
Data columns (total 14 columns):
 #   Column                   Non-Null Count   Dtype   
---  ------                   --------------   -----   
 0   Index                    660679 non-null  category
 1   Accident_Severity        660679 non-null  category
 2   Accident Date            660679 non-null  category
 3   Latitude                 660654 non-null  float64 
 4   Light_Conditions         660679 non-null  category
 5   District Area            660679 non-null  category
 6   Longitude                660653 non-null  float64 
 7   Number_of_Casualties     660679 non-null  int64   
 8   Number_of_Vehicles       660679 non-null  int64   
 9   Road_Surface_Conditions  659953 non-null  category
 10  Road_Type                656159 non-null  category
 11  Urban_or_Rural_Area      660664 non-null  category
 12  Weather_Conditions       646551 non-null  category
 13  Vehicle_Type             660679 non-null  ca

<h1>Filling in the Null Value</h1>
<hr>

In [9]:
Data.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                      25
Light_Conditions               0
District Area                  0
Longitude                     26
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions      726
Road_Type                   4520
Urban_or_Rural_Area           15
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

In [10]:
Data['Latitude'] = Data['Latitude'].fillna(Data['Latitude'].mean())
Data['Longitude'] = Data['Longitude'].fillna(Data['Longitude'].mean())
Data['Road_Surface_Conditions'] = Data['Road_Surface_Conditions'].fillna(Data['Road_Surface_Conditions'].mode()[0]) 
Data['Road_Type'] = Data['Road_Type'].fillna(Data['Road_Type'].mode()[0]) 
Data['Urban_or_Rural_Area'] = Data['Urban_or_Rural_Area'].fillna(Data['Urban_or_Rural_Area'].mode()[0]) 
Data['Weather_Conditions'] = Data['Weather_Conditions'].fillna(Data['Weather_Conditions'].mode()[0]) 

In [11]:
Data.isnull().sum()

Index                      0
Accident_Severity          0
Accident Date              0
Latitude                   0
Light_Conditions           0
District Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
dtype: int64

In [12]:
Data

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious      5/6/2019  51.506187   
1       200701BS65737           Serious      2/7/2019  51.495029   
2       200701BS66127           Serious    26-08-2019  51.517715   
3       200701BS66128           Serious    16-08-2019  51.495478   
4       200701BS66837            Slight      3/9/2019  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    18-02-2022  57.374005   
660675  201091NM01881            Slight    21-02-2022  57.232273   
660676  201091NM01935            Slight    23-02-2022  57.585044   
660677  201091NM01964           Serious    23-02-2022  57.214898   
660678  201091NM02142           Serious    28-02-2022  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2       Single carriageway               Urban     Fine no high winds   
3       Single carriageway               Urban     Fine no high winds   
4       Single carriageway               Urban     Fine no high winds   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  
0                                        Car  
1                                        Car  
2                      Taxi/Private hire car  
3       Bus or coach (17 or more pass seats)  
4                              Other vehicle  
...                                      ...  
660674                                   Car  
660675                                   Car  
660676                                   Car  
660677                 Motorcycle over 500cc  
660

<h1>EXPLORATORY DATA ANALYTICS(EDA)</h1>

<h1> 1. How many incidents have 5 or more casualties that include?</h1> 
<hr>

In [13]:
Data[Data['Number_of_Casualties'] > 4].value_counts().sum()

np.int64(6735)

<h1>INSIGHT</h1>
<hr>
<h3>There have been 6,735 accidents have 5 or more Casualties in each incident.</h3>

<h1>2. What are the common and most frequent accidents associated with road types?</h1>
<hr>

In [14]:
Data['Road_Type'].unique()

['Single carriageway', 'Dual carriageway', 'One way street', 'Roundabout', 'Slip road']
Categories (5, object): ['Dual carriageway', 'One way street', 'Roundabout', 'Single carriageway', 'Slip road']

In [15]:
Data[Data['Road_Type'] == 'Single carriageway'].value_counts().sum()

np.int64(496663)

In [16]:
Data[Data['Road_Type'] == 'Dual carriageway'].value_counts().sum()

np.int64(99424)

In [17]:
Data[Data['Road_Type'] == 'One way street'].value_counts().sum()

np.int64(13559)

In [18]:
Data[Data['Road_Type'] == 'Roundabout'].value_counts().sum()

np.int64(43992)

In [19]:
Data[Data['Road_Type'] == 'Slip road'].value_counts().sum()

np.int64(7041)

<h1>INSIGHT</h1>
<hr>
<h3>The common road had more accidents. It was a Single carriageway, because somehow, many cars were traveling at high speeds.</h3>

<h1>3. Which Area has many accidents?</h1>
<hr>

In [20]:
Data['District Area'].unique()

['Kensington and Chelsea', 'Westminster', 'Richmond upon Thames', 'Hammersmith and Fulham', 'Hounslow', ..., 'County Durham', 'Shropshire', 'Central Bedfordshire', 'Cornwall', 'Wiltshire']
Length: 422
Categories (422, object): ['Aberdeen City', 'Aberdeenshire', 'Adur', 'Allerdale', ..., 'Wycombe', 'Wyre', 'Wyre Forest', 'York']

In [21]:
Data[Data['District Area'] == 'Kensington and Chelsea'].value_counts().sum()

np.int64(2799)

In [22]:
Data[Data['District Area'] == 'Westminster'].value_counts().sum()

np.int64(5706)

In [23]:
Data[Data['District Area'] == 'Richmond upon Thames'].value_counts().sum()

np.int64(1607)

In [24]:
Data[Data['District Area'] == 'Hammersmith and Fulham'].value_counts().sum()

np.int64(2509)

In [25]:
Data[Data['District Area'] == 'Hounslow'].value_counts().sum()

np.int64(2979)

In [26]:
Data[Data['District Area'] == '...'].value_counts().sum()

np.int64(0)

In [27]:
Data[Data['District Area'] == 'County Durham'].value_counts().sum()

np.int64(2228)

In [28]:
Data[Data['District Area'] == 'Shropshire'].value_counts().sum()

np.int64(1343)

In [29]:
Data[Data['District Area'] == 'Central Bedfordshire'].value_counts().sum()

np.int64(1225)

In [30]:
Data[Data['District Area'] == 'Cornwall'].value_counts().sum()

np.int64(2606)

In [31]:
Data[Data['District Area'] == 'Wiltshire'].value_counts().sum()

np.int64(1590)

<h1>INSIGHT</h1>
<hr>
<h3> Based on all Areas, there are many accidents, but the Area of 'Westminster' has the highest accidents.</h3>

<h1>4. Get all the car incidents in Highland area</h1>
<hr>

In [32]:
Data[(Data['Vehicle_Type'] == 'car') & (Data['District Area'] == 'Highland')]

Empty DataFrame
Columns: [Index, Accident_Severity, Accident Date, Latitude, Light_Conditions, District Area, Longitude, Number_of_Casualties, Number_of_Vehicles, Road_Surface_Conditions, Road_Type, Urban_or_Rural_Area, Weather_Conditions, Vehicle_Type]
Index: []

<h1>INSIGHT</h1>
<hr>
<h3>We can see all the car accidents in highland area.</h3>

<h1>5. Get the number of Casualties greater than 2 but not exceeding 5</h1>
<hr>

In [33]:
Data_df = Data[Data['Number_of_Casualties'] > 1] 

In [34]:
Data_df[Data_df['Number_of_Casualties'] < 6].value_counts().sum()

np.int64(154026)

<h1>INSIGHT</h1>
<hr>
<h3>The number of accidents that have 2 to 5 casualties involved is '154,026'.</h3>

<h1>6. Get the Serious accidents that have 2 or more casualties involved</h1>
<hr>

In [35]:
Data[(Data['Accident_Severity'] == 'Serious') & (Data['Number_of_Casualties'] > 1)]

Index Accident_Severity Accident Date   Latitude  \
5       200701BS67159           Serious    18-09-2019  51.497750   
23      200701BS70002           Serious    22-01-2019  51.485674   
80      200701BS70066           Serious      3/2/2019  51.486026   
172     200701BS70174           Serious    16-04-2019  51.493411   
352     200701BS70366           Serious      9/7/2019  51.491519   
...               ...               ...           ...        ...   
660559  201091NJ10468           Serious    28-07-2022  58.440842   
660575  201091NJ13511           Serious    29-09-2022  58.289537   
660578  201091NJ14322           Serious    14-10-2022  58.606597   
660580  201091NJ14679           Serious    21-10-2022  58.099566   
660606  201091NK02539           Serious    18-05-2022  58.959892   

              Light_Conditions           District Area  Longitude  \
5                     Daylight  Kensington and Chelsea  -0.192556   
23                    Daylight  Kensington and Chelsea  -0.150543   
80                    Daylight  Kensington and Chelsea  -0.161476   
172                   Daylight  Kensington and Chelsea  -0.173712   
352                   Daylight  Kensington and Chelsea  -0.179262   
...                        ...                     ...        ...   
660559                Daylight                Highland  -3.093146   
660575  Darkness - no lighting                Highland  -3.342973   
660578  Darkness - no lighting                Highland  -3.344956   
660580                Daylight                Highland  -3.682299   
660606                Daylight          Orkney Islands  -3.005034   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
5                          2                   3                     Dry   
23                         2                   2                     Dry   
80                         2                   1                     Dry   
172                        2                   1                     Dry   
352                        2                   2                     Dry   
...                      ...                 ...                     ...   
660559                     2                   1             Wet or damp   
660575                     2                   1             Wet or damp   
660578                     2                   1             Wet or damp   
660580                     4                   2             Wet or damp   
660606                     2                   2                     Dry   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
5       Single carriageway               Urban     Fine no high winds   
23      Single carriageway               Urban     Fine no high winds   
80      Single carriageway               Urban     Fine no high winds   
172     Single carriageway               Urban     Fine no high winds   
352     Single carriageway               Urban     Fine no high winds   
...                    ...                 ...                    ...   
660559  Single carriageway               Rural                  Other   
660575    Dual carriageway               Rural  Raining no high winds   
660578  Single carriageway               Rural  Raining no high winds   
660580  Single carriageway               Rural  Raining no high winds   
660606  Single carriageway               Rural     Fine no high winds   

                               Vehicle_Type  
5                                       Car  
23                                      Car  
80                                      Car  
172                                     Car  
352     Van / Goods 3.5 tonnes mgw or under  
...                                     ...  
660559  Van / Goods 3.5 tonnes mgw or under  
660575  Van / Goods 3.5 tonnes mgw or under  
660578      Goods over 3.5t. and under 7.5t  
660580                                  Car  
660606                                  Car  

[23946 rows x 14 columns]

<h1>INSIGHT</h1>
<hr>
<h3>Here's the Serious accident that has 2 to many casualties involved.</h3>

<h1>7. Get all the accidents in the District of Cornwall that are caused by Wet road conditions </h1>
<hr


In [36]:
Data[(Data['District Area'] == 'Cornwall') & (Data['Road_Surface_Conditions'] == 'Wet or damp')]

Index Accident_Severity Accident Date   Latitude  \
484274  200950AC2B004            Slight      6/4/2021  50.180885   
484278  200950AC2B011            Slight      2/7/2021  50.194520   
484287  200950AC2C014            Slight    17-05-2021  50.224337   
484290  200950AC2C017            Slight      8/9/2021  50.214151   
484292  200950AC2C019            Slight    31-10-2021  50.222236   
...               ...               ...           ...        ...   
644156  201050GW3O008            Slight    14-05-2022  50.345577   
644166  201050GW3P007            Slight    20-07-2022  50.353734   
644168  201050GW3R007            Slight      7/8/2022  50.334243   
644171  201050GW3R010           Serious    17-11-2022  50.334272   
644778  201050JS2J001            Slight      7/2/2022  50.390923   

              Light_Conditions District Area  Longitude  Number_of_Casualties  \
484274                Daylight      Cornwall  -5.281875                     1   
484278  Darkness - no lighting      Cornwall  -5.308597                     3   
484287                Daylight      Cornwall  -5.288225                     1   
484290                Daylight      Cornwall  -5.298459                     1   
484292   Darkness - lights lit      Cornwall  -5.289342                     1   
...                        ...           ...        ...                   ...   
644156                Daylight      Cornwall  -4.741592                     2   
644166                Daylight      Cornwall  -4.700586                     1   
644168                Daylight      Cornwall  -4.636660                     1   
644171                Daylight      Cornwall  -4.639331                     1   
644778                Daylight      Cornwall  -4.421152                     1   

        Number_of_Vehicles Road_Surface_Conditions           Road_Type  \
484274                   2             Wet or damp  Single carriageway   
484278                   1             Wet or damp  Single carriageway   
484287                   2             Wet or damp  Single carriageway   
484290                   1             Wet or damp      One way street   
484292                   1             Wet or damp  Single carriageway   
...                    ...                     ...                 ...   
644156                   3             Wet or damp  Single carriageway   
644166                   1             Wet or damp      One way street   
644168                   1             Wet or damp  Single carriageway   
644171                   2             Wet or damp      One way street   
644778                   1             Wet or damp  Single carriageway   

       Urban_or_Rural_Area     Weather_Conditions  \
484274               Rural     Fine no high winds   
484278               Rural  Raining no high winds   
484287               Urban  Raining no high winds   
484290               Urban                  Other   
484292               Urban  Raining no high winds   
...                    ...                    ...   
644156               Rural  Raining no high winds   
644166               Rural  Raining no high winds   
644168               Rural     Fine no high winds   
644171               Rural     Fine no high winds   
644778               Rural     Fine no high winds   

                               Vehicle_Type  
484274                                  Car  
484278  Van / Goods 3.5 tonnes mgw or under  
484287                                  Car  
484290                                  Car  
484292                                  Car  
...                                     ...  
644156                                  Car  
644166  Van / Goods 3.5 tonnes mgw or under  
644168                                  Car  
644171                                  Car  
644778                                  Car  

[810 rows x 14 columns]

<h1>INSIGHT</h1>
<hr>
<h3>Here it is all the accidents in Cornwall that were caused by Wet road conditions.</h3>

<h1>8. Get all accidents in the Rural Area that have 1 casualty only</h1>

In [37]:
Data[(Data['Urban_or_Rural_Area'] == 'Rural') & (Data['Number_of_Casualties'])]

Index Accident_Severity Accident Date   Latitude  \
6101    200701JC30253            Slight    29-04-2019  51.616921   
6193    200701JC63842            Slight    23-06-2019  51.634035   
6223    200701JC64859            Slight    15-07-2019  51.616088   
6421    200701JC67563            Slight    17-10-2019  51.634365   
6441    200701JC67754            Slight     3/11/2019  51.615152   
...               ...               ...           ...        ...   
660672  201091NM01085           Serious    29-01-2022  57.170725   
660675  201091NM01881            Slight    21-02-2022  57.232273   
660676  201091NM01935            Slight    23-02-2022  57.585044   
660677  201091NM01964           Serious    23-02-2022  57.214898   
660678  201091NM02142           Serious    28-02-2022  57.575210   

                   Light_Conditions   District Area  Longitude  \
6101          Darkness - lights lit  Waltham Forest   0.010797   
6193                       Daylight  Waltham Forest   0.014444   
6223    Darkness - lighting unknown  Waltham Forest   0.012204   
6421         Darkness - no lighting  Waltham Forest   0.021395   
6441                       Daylight  Waltham Forest   0.014330   
...                             ...             ...        ...   
660672                     Daylight        Highland  -3.852421   
660675       Darkness - no lighting        Highland  -3.809281   
660676                     Daylight        Highland  -3.862727   
660677       Darkness - no lighting        Highland  -3.823997   
660678                     Daylight        Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
6101                       1                   1                     Dry   
6193                       1                   2             Wet or damp   
6223                       1                   3                     Dry   
6421                       1                   1             Wet or damp   
6441                       1                   2                     Dry   
...                      ...                 ...                     ...   
660672                     1                   2                    Snow   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
6101    Single carriageway               Rural     Fine no high winds   
6193    Single carriageway               Rural  Raining no high winds   
6223    Single carriageway               Rural     Fine no high winds   
6421    Single carriageway               Rural  Raining no high winds   
6441    Single carriageway               Rural     Fine no high winds   
...                    ...                 ...                    ...   
660672  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                 Vehicle_Type  
6101                      Car  
6193    Motorcycle over 500cc  
6223                      Car  
6421                      Car  
6441                      Car  
...                       ...  
660672                    Car  
660675                    Car  
660676                    Car  
660677  Motorcycle over 500cc  
660678                    Car  

[183817 rows x 14 columns]

<h1>INSIGHT</h1>
<hr>
<h3>We can see above all the accidents in the Rural area.</h3>

<h1>9. Retrieve all Car incidents that have involved three or more additional vehicles.</h1>


In [38]:
Data[(Data['Vehicle_Type'] == 'Car') & (Data['Number_of_Vehicles'] > 2)]

Index Accident_Severity Accident Date   Latitude  \
5       200701BS67159           Serious    18-09-2019  51.497750   
19      200701BS68237           Serious    26-11-2019  51.482988   
248     200701BS70255           Serious    16-05-2019  51.491632   
336     200701BS70350            Slight    22-06-2019  51.485178   
354     200701BS70368            Slight    29-06-2019  51.482076   
...               ...               ...           ...        ...   
660620  201091NK04948            Slight     4/10/2022  59.047179   
660641  201091NL04969            Slight      4/9/2022  60.151776   
660666  201091NM00770            Slight    21-01-2022  56.879446   
660668  201091NM00839            Slight    23-01-2022  57.054889   
660676  201091NM01935            Slight    23-02-2022  57.585044   

                   Light_Conditions           District Area  Longitude  \
5                          Daylight  Kensington and Chelsea  -0.192556   
19                         Daylight  Kensington and Chelsea  -0.174272   
248                        Daylight  Kensington and Chelsea  -0.192366   
336                        Daylight  Kensington and Chelsea  -0.187868   
354           Darkness - lights lit  Kensington and Chelsea  -0.173445   
...                             ...                     ...        ...   
660620                     Daylight          Orkney Islands  -3.196707   
660641        Darkness - lights lit        Shetland Islands  -1.148217   
660666  Darkness - lighting unknown                Highland  -4.255357   
660668                     Daylight                Highland  -4.130434   
660676                     Daylight                Highland  -3.862727   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
5                          2                   3                     Dry   
19                         1                   4                     Dry   
248                        1                   3                     Dry   
336                        1                   3                     Dry   
354                        1                   3             Wet or damp   
...                      ...                 ...                     ...   
660620                     6                   3                     Dry   
660641                     1                   3                     Dry   
660666                     2                   4                    Snow   
660668                     2                   3            Frost or ice   
660676                     1                   3            Frost or ice   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
5       Single carriageway               Urban     Fine no high winds   
19      Single carriageway               Urban     Fine no high winds   
248     Single carriageway               Urban     Fine no high winds   
336         One way street               Urban     Fine no high winds   
354     Single carriageway               Urban  Raining no high winds   
...                    ...                 ...                    ...   
660620          Roundabout               Rural     Fine no high winds   
660641  Single carriageway               Rural     Fine no high winds   
660666  Single carriageway               Rural   Snowing + high winds   
660668  Single carriageway               Rural                  Other   
660676  Single carriageway               Rural     Fine no high winds   

       Vehicle_Type  
5               Car  
19              Car  
248             Car  
336             Car  
354             Car  
...             ...  
660620          Car  
660641          Car  
660666          Car  
660668          Car  
660676          Car  

[51479 rows x 14 columns]

<h1>INSIGHT</h1>
<hr>
<h3> Here are the car accidents that involve more than 3 Vehicles. </h3>

<h1> 10. Which incident has the highest Vehicles involved?</h1>

In [39]:
Data['Number_of_Vehicles'].unique() 

array([ 2,  3,  4,  1,  5,  6, 14,  8,  7, 11, 10,  9, 16, 13, 12, 28, 15,
       32, 19])

In [40]:
Data[Data['Number_of_Vehicles'] == 32]

Index Accident_Severity Accident Date   Latitude  \
497581  2009559D05333            Slight    23-12-2021  50.752096   

       Light_Conditions District Area  Longitude  Number_of_Casualties  \
497581         Daylight       Purbeck  -2.254705                     5   

        Number_of_Vehicles Road_Surface_Conditions         Road_Type  \
497581                  32            Frost or ice  Dual carriageway   

       Urban_or_Rural_Area  Weather_Conditions Vehicle_Type  
497581               Rural  Fine no high winds          Car

<h1>INSIGHT</h1>
<hr>
<h3>The Vehicle involved in the incident is 32 Vehicles.</h3> 

<h1>11. Group the accidents by Accident_Severity and count the total number of cases in each category.</h1>
<hr>

In [41]:
Data.groupby(["Accident_Severity"]).size()

Accident_Severity
Fatal        8661
Serious     88217
Slight     563801
dtype: int64

<h1>INSIGHT</h1>
<hr>
<h3>We can see now which condition of the accidents has the highest count.</h3> 

<h1>12. Group accidents by District Area and find the district with the highest accident count, and what is the condition os the accidents.</h1>
<hr>

In [42]:
Data.groupby(['District Area', 'Accident_Severity']).size()

District Area  Accident_Severity
Aberdeen City  Fatal                  12
               Serious               239
               Slight               1072
Aberdeenshire  Fatal                  66
               Serious               463
                                    ... 
Wyre Forest    Serious               132
               Slight                815
York           Fatal                  21
               Serious               255
               Slight               1621
Length: 1266, dtype: int64

<h1>INSIGHT</h1>
<hr>
<h3>We can see now the accidents in each country/area have the highest count, and the condition of accidents.</h3> 

<h1>13. Group accidents by Light_Conditions and calculate the average casualties in each condition.</h1>
<hr>

In [43]:
Data.groupby(["Light_Conditions"])["Number_of_Casualties"].mean()

Light_Conditions
Darkness - lighting unknown    1.306447
Darkness - lights lit          1.369374
Darkness - lights unlit        1.331105
Darkness - no lighting         1.541657
Daylight                       1.340309
Name: Number_of_Casualties, dtype: float64

<h1>INSIGHT</h1>
<hr>
<h3>As we calculate what is the average of casualties of each incident.</h3> 

<h1>14. Group accidents by Road_Surface_Conditions and count the number of accidents in each.</h1>
<hr>

In [44]:
Data.groupby(["Road_Surface_Conditions"]).size()

Road_Surface_Conditions
Dry                     448547
Flood over 3cm. deep      1017
Frost or ice             18517
Snow                      5890
Wet or damp             186708
dtype: int64

<h1>INSIGHT</h1>
<hr>
<h3>By using groupby, we see each category of Road condition in each accident. </h3> 

<h1> 15. Group accidents by Vehicle_Type and find which type has the highest accident severity in each accident condition.</h1>
<hr>

In [45]:
Data.groupby(["Vehicle_Type", "Accident_Severity"]).size()

Vehicle_Type                           Accident_Severity
Agricultural vehicle                   Fatal                    21
                                       Serious                 282
                                       Slight                 1644
Bus or coach (17 or more pass seats)   Fatal                   325
                                       Serious                3373
                                       Slight                22180
Car                                    Fatal                  6577
                                       Serious               66461
                                       Slight               424954
Data missing or out of range           Fatal                     0
                                       Serious                   0
                                       Slight                    6
Goods 7.5 tonnes mgw and over          Fatal                   216
                                       Serious                2321
     

<h1>INSIGHT</h1>
<hr>
<h3>The highest Serious accident condition is a Car that has 66461 incidents.
The highest Fatal accident condition is a Car that has 6577 incidents.
The highest Slight accident condition is a Car that has 424954 incidents.</h3> 

<h1>16. Correlation between number of casualties and number of vehicles</h1>
<hr>

In [46]:
 Data["Number_of_Casualties"].corr(Data["Number_of_Vehicles"])

np.float64(0.2288888612692756)

<h1>INSIGHT</h1>
<hr>
<h3>there is no correlation for Number_of_Casualties and Number_of_Vehicles.</h3> 

<h1>17. Correlation between casualties and latitude </h1>
<hr>

In [47]:
Data["Number_of_Casualties"].corr(Data["Latitude"])

np.float64(0.032200686625906395)

<h1>INSIGHT</h1>
<hr>
<h3>there was no relation for Number_of_Casualties and Latitude. </h3> 

<h1>18. Correlation between casualties and longitude</h1>
<hr>

In [48]:
Data["Number_of_Casualties"].corr(Data["Longitude"])

np.float64(-0.04040564578845451)

<h1>INSIGHT</h1>
<hr>
<h3>there was no relation for Number_of_Casualties and Longitude.  </h3> 

<h1>19. Which vehicle type is most frequently involved in accidents?</h1>
<hr>

In [49]:
Data["Vehicle_Type"].value_counts()

Vehicle_Type
Car                                      497992
Van / Goods 3.5 tonnes mgw or under       34160
Bus or coach (17 or more pass seats)      25878
Motorcycle over 500cc                     25657
Goods 7.5 tonnes mgw and over             17307
Motorcycle 125cc and under                15269
Taxi/Private hire car                     13294
Motorcycle over 125cc and up to 500cc      7656
Motorcycle 50cc and under                  7603
Goods over 3.5t. and under 7.5t            6096
Other vehicle                              5637
Minibus (8 - 16 passenger seats)           1976
Agricultural vehicle                       1947
Pedal cycle                                 197
Data missing or out of range                  6
Ridden horse                                  4
Name: count, dtype: int64

<h1>INSIGHT</h1>
<hr>
<h3>there was The most frequently involved in accidents is Car that has 497992 accidents. </h3> 

<h1>20. Which weather condition had the least number of accidents recorded?</h1>
<hr>

In [50]:
Data["Weather_Conditions"].value_counts()

Weather_Conditions
Fine no high winds       535013
Raining no high winds     79696
Other                     17150
Raining + high winds       9615
Fine + high winds          8554
Snowing no high winds      6238
Fog or mist                3528
Snowing + high winds        885
Name: count, dtype: int64

<h1>INSIGHT</h1>
<hr>
<h3>The least number of accidents in weather condition is Snowing + high winds that have 885 accidents.</h3> 